server/src/test/java/org/elasticsearch/snapshots/SnapshotsServiceTests.java

```diff
@@ -0,0 +1,526 @@
+    }
+
+    private static ClusterChangedEvent changeEventForNode(ClusterChangedEvent event, DiscoveryNode node) {
+        return new ClusterChangedEvent(event.source(), stateForNode(event.state(), node), stateForNode(event.previousState(), node));
+    }
+
+    private final class TestClusterNodes {
+
+        // LinkedHashMap so we have deterministic ordering when iterating over the map in tests
+        private final Map<String, TestClusterNode> nodes = new LinkedHashMap<>();
+
+        TestClusterNodes(int masterNodes, int dataNodes) {
+            for (int i = 0; i < masterNodes; ++i) {
+                nodes.computeIfAbsent("node" + i, nodeName -> {
+                    try {
+                        return SnapshotsServiceTests.this.newMasterNode(nodeName);
+                    } catch (IOException e) {
+                        throw new UncheckedIOException(e);
+                    }
+                });
+            }
+            for (int i = masterNodes; i < dataNodes + masterNodes; ++i) {
+                nodes.computeIfAbsent("node" + i, nodeName -> {
+                    try {
+                        return SnapshotsServiceTests.this.newDataNode(nodeName);
+                    } catch (IOException e) {
+                        throw new UncheckedIOException(e);
+                    }
+                });
+            }
+        }
+
+        /**
+         * Builds a {@link DiscoveryNodes} instance that has one master eligible node set as its master
+         * by random.
+         * @return DiscoveryNodes with set master node
+         */
+        public DiscoveryNodes randomDiscoveryNodes() {
+            DiscoveryNodes.Builder builder = DiscoveryNodes.builder();
+            nodes.values().forEach(node -> builder.add(node.node));
+            String masterId = randomFrom(nodes.values().stream().map(node -> node.node).filter(DiscoveryNode::isMasterNode)
+                .map(DiscoveryNode::getId)
+                .collect(Collectors.toList()));
+            return builder.localNodeId(masterId).masterNodeId(masterId).build();
+        }
+
+        /**
+         * Returns the {@link TestClusterNode} for the master node in the given {@link ClusterState}.
+         * @param state ClusterState
+         * @return Master Node
+         */
+        public TestClusterNode currentMaster(ClusterState state) {
+            TestClusterNode master = nodes.get(state.nodes().getMasterNode().getName());
+            assertNotNull(master);
+            assertTrue(master.node.isMasterNode());
+            return master;
+        }
+    }
+
+    private final class TestClusterNode {
+
+        private final DeterministicTaskQueue deterministicTaskQueue;
+
+        private final TransportService transportService;
+
+        private final ClusterService clusterService;
+
+        private final RepositoriesService repositoriesService = mock(RepositoriesService.class);
```

Comment:
Just mocked this one out since it's not that relevant for the test (we really only need it to return the repository, that's the only call we make to it) and just a bunch of code to get up and running.